File: spring-boot-starter-breuninger-jobs/src/main/java/com/breuninger/boot/jobs/configuration/JobsProperties.java
package com.breuninger.boot.jobs.configuration;

import java.util.HashMap;
import java.util.Map;

import javax.validation.Valid;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;

import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.validation.annotation.Validated;

import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
@ConfigurationProperties(prefix = "breuninger.jobs")
@Validated
public class JobsProperties {

  private boolean externalTrigger = true;
  private @Min(1) int threadCount = 10;
  private @Valid Cleanup cleanup = new Cleanup();
  private @Valid Status status = new Status();

  @Getter
  @Setter
  public static class Cleanup {

    private @Min(1) int numberOfJobsToKeep = 100;
    private @Min(1) int numberOfSkippedJobsToKeep = 10;
    private @Min(1) int markDeadAfter = 30;
  }

  @Getter
  @Setter
  public static class Status {

    private boolean enabled = true;
    private @NotNull Map<String, String> calculator = new HashMap<>();

    public void setCalculator(final Map<String, String> calculator) {
      final Map<String, String> normalized = new HashMap<>();
      calculator.forEach((key, value) -> normalized.put(key.toLowerCase().replace(" ", "-"), value));
      this.calculator = normalized;
    }
  }
}

File: jobs/src/main/kotlin/com/breuninger/boot/jobs/autoconfigure/JobsProperties.kt
package com.breuninger.boot.jobs.autoconfigure

import org.springframework.boot.context.properties.ConfigurationProperties
import org.springframework.validation.annotation.Validated
import javax.validation.constraints.Min

@ConfigurationProperties(prefix = "breuni.jobs")
@Validated
class JobsProperties {

  var enabled = false

  @field:Min(1)
  var threadCount = 10

  val mongo = Mongo()

  class Mongo {

    var enabled = false
  }

  val cleanup = Cleanup()

  class Cleanup {

    @field:Min(1)
    var numberOfJobExecutionsToKeep = 100

    @field:Min(1)
    var numberOfSkippedJobExecutionsToKeep = 10

    @field:Min(30)
    var killDeadJobExecutionsAfterSeconds = 30
  }
}
